include/GenSync/Syncs/BloomFilter.h

```diff
@@ -0,0 +1,231 @@
+
+    // default constructor
+    BloomFilter();
+
+    // default destructor
+    ~BloomFilter();
+
+    /**
+     * Getter for size of BloomFilter, specifically its length in bits.
+     * @return size_t The length of the BloomFilter's bit string
+     */
+    size_t getSize();
+
+    /**
+     * Getter for number of hash functions used by BloomFilter.
+     * @return int The number of hash functions used by BloomFilter per element
+     */
+    int getNumHashes();
+
+    /**
+     * Getter for BloomFilter's bits in form of vector<bool>.
+     * @return vector<bool> The bit string
+     */
+    vector<bool> getBits();
+
+    /**
+     * Calculates BloomFilter's rate of false positives.
+     * @param numExpElems The expected number of elements stored in the BloomFilter
+     * @return float The false positive probability/rate
+     */
+    float getFalsePosProb(size_t numExpElems);
+
+    /**
+     * Insert an element into BloomFilter.
+     * @param value The element to be added to BloomFilter
+     */
+    void insert(ZZ value);
+    
+    /**
+     * Insert a set of elements into BloomFilter.
+     * @param tarSet target set to be added to BloomFilter
+     */
+    void insert(multiset<shared_ptr<DataObject>> tarSet);
+
+    /**
+     * Probabilistically determine whether an element is in BloomFilter's bit string.
+     * @param value The element to be determined if in BloomFilter
+     * @return true iff element is probably in BloomFilter, false if definitely not.
+     */
+    bool exist(ZZ value);
+
+    /**
+     * Convert BloomFilter to a readable string.
+     * @return string
+     */
+    string toString() const;
+
+    /**
+     * Convert BloomFilter to a ZZ.
+     * @return ZZ
+     */
+    ZZ toZZ();
+
+    /**
+     * Convert ZZ to BloomFilter.
+     * @param ZZ The ZZ to be converted
+     * @return BloomFilter
+     */
+    BloomFilter ZZtoBF(ZZ val);
+
+protected:
+    // constructors should not be used, only builder pattern should be accessible
+
+    /**
+     * Constructs a BloomFilter object with bit string's size equal to size.
+     * @param size The size of BloomFilter, specifically its length in bits
+     * @param nHash The number of hash functions BloomFilter will use for each element insertion
+     */
+    BloomFilter(size_t size, int nHash);
+
+    /**
+     * Constructs a BloomFilter object from false positive rate as input.
+     * @param numExpElems The expected number of elements in the BloomFilter
+     * @param falsePosProb The rate of false positives
+     */
+    BloomFilter(size_t numExpElems, float falsePosProb);
+
+    /** 
+     * Returns the kk-th hash of the ZZ value.
+     * @param value The value being hashed
+     * @param kk The current iteration of hash(value)
+     * @return hash(hash(value) + hash(kk))
+     */
+    static hash_t _hash(const ZZ& value, long kk);
+
+    // all bits in BloomFilter
+    vector<bool> bits;
+    
+    // Length of the BloomFilter in bits
+    size_t bfSize;

```

Comment:
can't this be inferred from the "bits" variable?  It's not a good idea to have independent variables for derived values, because that introduces the potential for a mismatch.